File: adapters/kotlin/src/main/kotlin/Rfc8949.kt
import java.math.BigInteger

class EncodeException(message: String) : Exception(message)

private val TWO_POW_64: BigInteger = BigInteger.ONE.shiftLeft(64)

private fun bigIntToBytes(v: BigInteger, width: Int): ByteArray {
    val raw = v.toByteArray()
    val unsigned = if (raw.size > 1 && raw[0] == 0.toByte()) raw.copyOfRange(1, raw.size) else raw
    val out = ByteArray(width)
    unsigned.copyInto(out, destinationOffset = width - unsigned.size)
    return out
}

private fun stripLeadingZeros(bytes: ByteArray): ByteArray {
    var start = 0
    while (start < bytes.size - 1 && bytes[start] == 0.toByte()) start++
    return bytes.copyOfRange(start, bytes.size)
}

fun encodeHead(majorType: Int, arg: BigInteger): ByteArray {
    val mt = majorType shl 5
    return when {
        arg < BigInteger.valueOf(24) -> byteArrayOf((mt or arg.toInt()).toByte())
        arg <= BigInteger.valueOf(0xff) -> byteArrayOf((mt or 24).toByte()) + bigIntToBytes(arg, 1)
        arg <= BigInteger.valueOf(0xffff) -> byteArrayOf((mt or 25).toByte()) + bigIntToBytes(arg, 2)
        arg <= BigInteger.valueOf(0xffffffffL) -> byteArrayOf((mt or 26).toByte()) + bigIntToBytes(arg, 4)
        else -> byteArrayOf((mt or 27).toByte()) + bigIntToBytes(arg, 8)
    }
}

// Not private: reused by Dcbor.kt (same shortest-int rule applies to both profiles).
fun encodeInt(value: String): ByteArray {
    val n = try {
        BigInteger(value)
    } catch (e: Exception) {
        throw EncodeException("int: invalid integer literal \"$value\": ${e.message}")
    }
    return if (n >= BigInteger.ZERO) {
        if (n >= TWO_POW_64) throw EncodeException("int: $n exceeds native range")
        encodeHead(0, n)
    } else {
        val arg = BigInteger.valueOf(-1).subtract(n)
        if (arg >= TWO_POW_64) throw EncodeException("int: $n exceeds native range")
        encodeHead(1, arg)
    }
}

// Not private: reused by Dcbor.kt.
fun doubleToBeBytes(f: Double): ByteArray {
    val bits = java.lang.Double.doubleToRawLongBits(f)
    val out = ByteArray(8)
    for (i in 0 until 8) out[i] = ((bits ushr (56 - 8 * i)) and 0xffL).toByte()
    return out
}

private fun parseFloatLiteral(literal: String): Double =
    literal.toDoubleOrNull() ?: throw EncodeException("float: invalid literal \"$literal\"")

private fun encodeFloat(width: String, literal: String): ByteArray {
    val f = parseFloatLiteral(literal)
    if (f.isNaN()) {
        // P6: NaN always uses the single canonical f16 payload, regardless of width.
        return byteArrayOf(0xf9.toByte(), 0x7e, 0x00)
    }
    if (width == "auto") {
        tryF16(f)?.let { return byteArrayOf(0xf9.toByte()) + it }
        tryF32(f)?.let { return byteArrayOf(0xfa.toByte()) + it }
        return byteArrayOf(0xfb.toByte()) + doubleToBeBytes(f)
    }
    return when (width) {
        "f16" -> byteArrayOf(0xf9.toByte()) + (tryF16(f)
            ?: throw EncodeException("float: $f cannot round-trip at requested width f16"))
        "f32" -> byteArrayOf(0xfa.toByte()) + (tryF32(f)
            ?: throw EncodeException("float: $f cannot round-trip at requested width f32"))
        "f64" -> byteArrayOf(0xfb.toByte()) + doubleToBeBytes(f)
        else -> throw EncodeException("float: unknown width \"$width\"")
    }
}

// Shared with the dcbor profile: bignum tag+magnitude rules are identical in
// both profiles, so both encoders build on this single validated computation.
fun bignumTagAndBytes(sign: String, value: String): Pair<Int, ByteArray> {
    val magnitude = try {
        BigInteger(value)
    } catch (e: Exception) {
        throw EncodeException("bignum: invalid magnitude \"$value\": ${e.message}")
    }
    if (magnitude < TWO_POW_64) {
        val tag = if (sign == "positive") 2 else 3
        throw EncodeException(
            "bignum magnitude $magnitude fits in a native CBOR integer and must not be encoded as tag $tag (SPEC.md bignum rule)"
        )
    }
    val rawInt = when (sign) {
        "positive" -> magnitude
        "negative" -> magnitude.subtract(BigInteger.ONE)
        else -> throw EncodeException("bignum: unknown sign \"$sign\"")
    }
    val tag = if (sign == "positive") 2 else 3
    val bytes = stripLeadingZeros(rawInt.toByteArray())
    return tag to bytes
}

private fun encodeBignum(sign: String, value: String): ByteArray {
    val (tag, bytes) = bignumTagAndBytes(sign, value)
    var out = encodeHead(6, BigInteger.valueOf(tag.toLong()))
    out += encodeHead(2, BigInteger.valueOf(bytes.size.toLong()))
    out += bytes
    return out
}

// Not private: reused by Dcbor.kt (same bytewise map-key sort rule applies to both profiles).
fun compareBytesUnsigned(a: ByteArray, b: ByteArray): Int {
    val len = minOf(a.size, b.size)
    for (i in 0 until len) {
        val ai = a[i].toInt() and 0xff
        val bi = b[i].toInt() and 0xff
        if (ai != bi) return ai - bi
    }
    return a.size - b.size
}

fun encodeRfc8949(value: LogicalValue): ByteArray = when (value) {
    is LogicalValue.Int -> encodeInt(value.value)
    is LogicalValue.Text -> {
        val bytes = value.value.toByteArray(Charsets.UTF_8)
        encodeHead(3, BigInteger.valueOf(bytes.size.toLong())) + bytes
    }
    is LogicalValue.Bytes -> {
        val bytes = hexDecode(value.value)
        encodeHead(2, BigInteger.valueOf(bytes.size.toLong())) + bytes
    }
    is LogicalValue.Bool -> byteArrayOf(if (value.value) 0xf5.toByte() else 0xf4.toByte())
    LogicalValue.Null -> byteArrayOf(0xf6.toByte())
    is LogicalValue.Array -> {
        var out = encodeHead(4, BigInteger.valueOf(value.items.size.toLong()))
        for (item in value.items) out += encodeRfc8949(item)
        out
    }
    is LogicalValue.Float -> encodeFloat(value.width, value.value)
    is LogicalValue.Map -> {
        val encoded = value.entries.map { (k, v) -> encodeRfc8949(k) to encodeRfc8949(v) }
        val sorted = encoded.sortedWith { a, b -> compareBytesUnsigned(a.first, b.first) }
        var out = encodeHead(5, BigInteger.valueOf(value.entries.size.toLong()))
        for ((k, v) in sorted) {
            out += k
            out += v
        }
        out
    }
    is LogicalValue.Tag -> encodeHead(6, BigInteger(value.tag.toString())) + encodeRfc8949(value.value)
    is LogicalValue.Bignum -> encodeBignum(value.sign, value.value)
}

File: adapters/typescript/src/rfc8949.ts
/**
 * RFC 8949 §4.2.1 deterministic-encoding profile ("P" rules in SPEC.md).
 * Faithful port of adapters/kotlin/src/main/kotlin/Rfc8949.kt.
 */

import { hexDecode } from "./util.ts";
import { tryF16, tryF32 } from "./float16.ts";
import type { LogicalValue } from "./logicalValue.ts";

export class EncodeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "EncodeError";
  }
}

export const TWO_POW_64 = 1n << 64n;

/** Convert a non-negative bigint magnitude to a big-endian byte array of exactly `width` bytes. */
function bigIntToBytes(v: bigint, width: number): Uint8Array {
  const out = new Uint8Array(width);
  let n = v;
  for (let i = width - 1; i >= 0; i--) {
    out[i] = Number(n & 0xffn);
    n >>= 8n;
  }
  return out;
}

/** Convert a non-negative bigint to the minimal big-endian byte array (at least 1 byte). */
function bigIntToMinimalBytes(v: bigint): Uint8Array {
  if (v === 0n) return new Uint8Array([0]);
  let hex = v.toString(16);
  if (hex.length % 2 !== 0) hex = "0" + hex;
  return hexDecode(hex);
}

export function concatBytes(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((sum, p) => sum + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

export function encodeHead(majorType: number, arg: bigint): Uint8Array {
  const mt = majorType << 5;
  if (arg < 24n) {
    return new Uint8Array([mt | Number(arg)]);
  } else if (arg <= 0xffn) {
    return concatBytes(new Uint8Array([mt | 24]), bigIntToBytes(arg, 1));
  } else if (arg <= 0xffffn) {
    return concatBytes(new Uint8Array([mt | 25]), bigIntToBytes(arg, 2));
  } else if (arg <= 0xffffffffn) {
    return concatBytes(new Uint8Array([mt | 26]), bigIntToBytes(arg, 4));
  } else {
    return concatBytes(new Uint8Array([mt | 27]), bigIntToBytes(arg, 8));
  }
}

/** Not exported-private: reused by Dcbor.ts (same shortest-int rule applies to both profiles). */
export function encodeInt(value: string): Uint8Array {
  let n: bigint;
  try {
    n = BigInt(value);
  } catch (e) {
    throw new EncodeError(
      `int: invalid integer literal "${value}": ${e instanceof Error ? e.message : String(e)}`,
    );
  }
  if (n >= 0n) {
    if (n >= TWO_POW_64) throw new EncodeError(`int: ${n} exceeds native range`);
    return encodeHead(0, n);
  } else {
    const arg = -1n - n;
    if (arg >= TWO_POW_64) throw new EncodeError(`int: ${n} exceeds native range`);
    return encodeHead(1, arg);
  }
}

/** Reused by Dcbor.ts. */
export function doubleToBeBytes(f: number): Uint8Array {
  const view = new DataView(new ArrayBuffer(8));
  view.setFloat64(0, f, false);
  return new Uint8Array(view.buffer);
}

function parseFloatLiteral(literal: string): number {
  const f = Number(literal);
  if (Number.isNaN(f) && literal.trim().toLowerCase() !== "nan") {
    throw new EncodeError(`float: invalid literal "${literal}"`);
  }
  return f;
}

function encodeFloat(width: "auto" | "f16" | "f32" | "f64", literal: string): Uint8Array {
  const f = parseFloatLiteral(literal);
  if (Number.isNaN(f)) {
    // P6: NaN always uses the single canonical f16 payload, regardless of width.
    return new Uint8Array([0xf9, 0x7e, 0x00]);
  }
  if (width === "auto") {
    const f16 = tryF16(f);
    if (f16) return concatBytes(new Uint8Array([0xf9]), f16);
    const f32 = tryF32(f);
    if (f32) return concatBytes(new Uint8Array([0xfa]), f32);
    return concatBytes(new Uint8Array([0xfb]), doubleToBeBytes(f));
  }
  switch (width) {
    case "f16": {
      const bytes = tryF16(f);
      if (!bytes) throw new EncodeError(`float: ${f} cannot round-trip at requested width f16`);
      return concatBytes(new Uint8Array([0xf9]), bytes);
    }
    case "f32": {
      const bytes = tryF32(f);
      if (!bytes) throw new EncodeError(`float: ${f} cannot round-trip at requested width f32`);
      return concatBytes(new Uint8Array([0xfa]), bytes);
    }
    case "f64":
      return concatBytes(new Uint8Array([0xfb]), doubleToBeBytes(f));
    default:
      throw new EncodeError(`float: unknown width "${width}"`);
  }
}

/**
 * Shared with the dcbor profile: bignum tag+magnitude rules are identical in
 * both profiles, so both encoders build on this single validated computation.
 */
export function bignumTagAndBytes(
  sign: "positive" | "negative",
  value: string,
): { tag: number; bytes: Uint8Array } {
  let magnitude: bigint;
  try {
    magnitude = BigInt(value);
  } catch (e) {
    throw new EncodeError(
      `bignum: invalid magnitude "${value}": ${e instanceof Error ? e.message : String(e)}`,
    );
  }
  const tag = sign === "positive" ? 2 : 3;
  if (magnitude < TWO_POW_64) {
    throw new EncodeError(
      `bignum magnitude ${magnitude} fits in a native CBOR integer and must not be encoded as tag ${tag} (SPEC.md bignum rule)`,
    );
  }
  let rawInt: bigint;
  if (sign === "positive") {
    rawInt = magnitude;
  } else if (sign === "negative") {
    rawInt = magnitude - 1n;
  } else {
    throw new EncodeError(`bignum: unknown sign "${sign}"`);
  }
  const bytes = bigIntToMinimalBytes(rawInt);
  return { tag, bytes };
}

function encodeBignum(sign: "positive" | "negative", value: string): Uint8Array {
  const { tag, bytes } = bignumTagAndBytes(sign, value);
  return concatBytes(
    encodeHead(6, BigInt(tag)),
    encodeHead(2, BigInt(bytes.length)),
    bytes,
  );
}

/** Reused by Dcbor.ts (same bytewise map-key sort rule applies to both profiles). */
export function compareBytesUnsigned(a: Uint8Array, b: Uint8Array): number {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const ai = a[i];
    const bi = b[i];
    if (ai !== bi) return ai - bi;
  }
  return a.length - b.length;
}

export function encodeRfc8949(value: LogicalValue): Uint8Array {
  switch (value.type) {
    case "int":
      return encodeInt(value.value);
    case "text": {
      const bytes = new TextEncoder().encode(value.value);
      return concatBytes(encodeHead(3, BigInt(bytes.length)), bytes);
    }
    case "bytes": {
      const bytes = hexDecode(value.value);
      return concatBytes(encodeHead(2, BigInt(bytes.length)), bytes);
    }
    case "bool":
      return new Uint8Array([value.value ? 0xf5 : 0xf4]);
    case "null":
      return new Uint8Array([0xf6]);
    case "array": {
      const parts = [encodeHead(4, BigInt(value.items.length))];
      for (const item of value.items) parts.push(encodeRfc8949(item));
      return concatBytes(...parts);
    }
    case "float":
      return encodeFloat(value.width, value.value);
    case "map": {
      const encoded = value.entries.map(
        ([k, v]) => [encodeRfc8949(k), encodeRfc8949(v)] as [Uint8Array, Uint8Array],
      );
      const sorted = encoded
        .map((pair, index) => ({ pair, index }))
        .sort((a, b) => {
          const cmp = compareBytesUnsigned(a.pair[0], b.pair[0]);
          return cmp !== 0 ? cmp : a.index - b.index;
        })
        .map((x) => x.pair);
      const parts = [encodeHead(5, BigInt(value.entries.length))];
      for (const [k, v] of sorted) {
        parts.push(k);
        parts.push(v);
      }
      return concatBytes(...parts);
    }
    case "tag":
      return concatBytes(encodeHead(6, value.tag), encodeRfc8949(value.value));
    case "bignum":
      return encodeBignum(value.sign, value.value);
  }
}
